dash/development/base_component.py

```diff
@@ -132,13 +134,26 @@ def __init__(self, **kwargs):
                     f"\nAllowed arguments: {allowed_args}"
                 )
 
-            if k != "children" and isinstance(v, Component):
-                raise TypeError(
-                    error_string_prefix
-                    + " detected a Component for a prop other than `children`\n"
-                    + "Did you forget to wrap multiple `children` in an array?\n"
-                    + f"Prop {k} has value {v!r}\n"
-                )
+            if k not in ("children", "id", "style", "className") and not k_in_wildcards:
+                if isinstance(v, Component):
+                    self._children_props.append(k)
+                elif isinstance(v, dict):
+                    for key, value in v.items():
+                        if isinstance(value, Component):
+                            self._children_props.append(k + "." + key)
+                elif hasattr(v, "__iter__"):
+                    for item in v:
+                        if isinstance(item, Component):
+                            self._children_props.append(k)
+                            break
+                        if isinstance(item, dict):
+                            found = False
+                            for key, value in item.items():
+                                if isinstance(value, Component):
+                                    self._children_props.append("[]" + k + "." + key)
+                                    found = True
+                            if found:
+                                break
```

Comment:
I don't know if calculating `_children_props` at component instantiation is the right way - I was envisioning we'd do this either while building the components, flag all the props and prop parts that are allowed to be nodes... or even somehow have the renderer extract this info from the component `propTypes` at run time though the `PropTypes` package really doesn't seem to want you to do that...

Anyway the two problems I see with this:
- if you instantiate a component with only strings in these props, then a callback later provides components directly to the props, those props won't be in `_children_props`
- this requires iterating through potentially enormous props, like table data, even if we have no intention of allowing components there.
- and one due to your specific implementation (that we could fix, but as I said I think a bigger change is needed): if a list of dicts allows components in two different keys, but the first item in the list to have a component only has it on one of those keys, the other key won't be in `_children_props`

I'd love if we could handle this entirely in the renderer, so the only back-end update would be removing the . I suppose a real simple way would be to require manually enumerating component props in the js source code - like `Checklist.childrenProps = ['[]options.label']`. Not very 🌴 , but we could probably write something into the component generator to at least throw an error if this isn't done correctly?

Reply:
I added the node paths from the metadata into a class attribute, the renderer now wrap the props all the time, works fine but there is a little extra overhead for the union types. I was thinking maybe add a symbol in the key like `?` so it only performs a check in those cases?

Reply:
It also adds the children_props to all the components json, the additional data could be pretty significant for repetitive info. I think we should supply that info somewhere else.

Reply:
Right, the ideal would still be to provide these once per component class - which is why I suggested putting it directly into the JS source, but I still don't see a way to do that without manual duplication. But what about appending it to the entry point bundle when it's served, in [`serve_component_suites`](https://github.com/plotly/dash/blob/07be9e2d0c060b929540aee88457bba717123a66/dash/dash.py#L810)? Could be as simple as constructing a string like:
```
{namespace}.{component_name}.childrenProps = {repr(children_props)};
{namespace}.{component_name2}.childrenProps = {repr(children_props2)};
...
```
and appending it to the served file - assuming we can figure out which one is the entry point bundle. A little hacky perhaps... but maybe workable?

Reply:
I opted to use the component meta & registry to gather the children props, they are then served in the configs json script tag on the index. The TreeContainer has a prop `_dashprivate_config` so it's easy to get from there and extracted the configs from the store to use in `crawlLayout`. I think it works good now.